### DISCUSSION TITLE
OpenCL platform ID is empty; fsycl-target: spir64_x86_64

### DISCUSSION THREAD

**Discussion by dosarudaniel** [intel/llvm#3701] [Q&A] [answered]
When I am trying to compile this [sample code](https://github.com/dosarudaniel/oneAdePT/blob/62a0612372291d97877920767c3d6c394270c0f7/tests/test12.cpp) (`make test12`) , with this coammand:
```
clang++ -O3 -DNDEBUG CMakeFiles/test12.dir/test12.cpp.o -o test12  -Wl,-rpath,/usr/local/lib /home/dadosaru/local/lib/libvecgeom.a /home/dadosaru/local/lib/libvecgeomcuda_static.a /home/dadosaru/local/lib/libvgdml.a -fsycl -fsycl-unnamed-lambda \
-fsycl-targets=spir64_x86_64-unknown-unknown-sycldevice,nvptx64-nvidia-cuda-sycldevice \
-Xsycl-target-backend=nvptx64-nvidia-cuda-sycldevice \
--cuda-gpu-arch=sm_61 \
-std=c++20 -stdlib=libstdc++ -Wno-unknown-cuda-version -D_ONEADEPT_ -DNDEBUG -DROOT_NO_INT128 -DG4HepEm_CUDA_BUILD -I/home/dadosaru/oneAdePT/../local/include /home/dadosaru/local/lib/libvecgeom.a -ldl -lrt -lpthread /usr/local/lib/libxerces-c.so
```

I get this error:
```
OpenCL platform ID is empty
OpenCL platform name is empty
Failed to find any of these OpenCL platforms:
  Intel(R) OpenCL
  Intel(R) CPU Runtime for OpenCL(TM) Applications
llvm-foreach: 
clang-13: error: x86_64 compiler command failed with exit code 1 (use -v to see invocation)
```

Is something wrong with my environment or is it a bug?   

If I remove the `spir64_x86_64-unknown-unknown-sycldevice` from `fsycl-targets` it compile and link just fine.

**Comment by alexbatashev** [comment, marked as answer]
You're trying to build your image for x86_64 platform in AOT mode. This requires you to have a fully set up OpenCL CPU device (see [Get Started Guide](https://intel.github.io/llvm-docs/GetStartedGuide.html)). The way it works is the compiler will produce a SPIR-V binary, and feed it to `opencl-aot` tool which will essentially call some OpenCL APIs to produce the x64 binary executable, which can later be consumed by OpenCL runtime directly. If you want to simply build platform-agnostic SPIR-V, you should use `spir64-unknown-unknown-sycldevice` target.

**Reply by dosarudaniel** [reply]
Thanks!
I've followed [these steps](https://intel.github.io/llvm-docs/GetStartedGuide.html#install-low-level-runtime) as you suggested to Install the low level runtime and it worked.

**Comment by Gabriel-MarcoSilva** [comment]
I need help! I have a similar problem, but when I run make fpga_emu:

OpenCL platform ID is empty
OpenCL platform name is empty
Failed to find any of these OpenCL platforms:
Intel(R) FPGA Emulation Platform for OpenCL(TM)
Intel(R) FPGA Emulation Platform for OpenCL(TM) (preview)
llvm-foreach:
icpx: error: fpga compiler command failed with exit code 14 (use -v to see invocation)
make[3]: *** [CMakeFiles/fpga_emu.dir/build.make:97: vector_add.fpga_emu] Error 14
make[2]: *** [CMakeFiles/Makefile2:85: CMakeFiles/fpga_emu.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:92: CMakeFiles/fpga_emu.dir/rule] Error 2
make: *** [Makefile:124: fpga_emu] Error 2
